LLama/Native/NativeLogConfig.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Extensions.Logging;
+
+namespace LLama.Native;
+
+/// <summary>
+/// Configure llama.cpp logging
+/// </summary>
+public static class NativeLogConfig
+{
+    /// <summary>
+    /// Callback from llama.cpp with log messages
+    /// </summary>
+    /// <param name="level"></param>
+    /// <param name="message"></param>
+    public delegate void LLamaLogCallback(LLamaLogLevel level, string message);
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logCallback"></param>
+    [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl, EntryPoint = "llama_log_set")]
+    private static extern void native_llama_log_set(LLamaLogCallback? logCallback);
+
+    /// <summary>
+    /// A GC handle for the current log callback to ensure the callback is not collected
+    /// </summary>
+    private static GCHandle? _currentLogCallbackHandle;
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logCallback"></param>
+#pragma warning disable IDE1006 // Naming Styles (name imitates llama.cpp)
+    public static void llama_log_set(LLamaLogCallback? logCallback)
+#pragma warning restore IDE1006 // Naming Styles
+    {
+        if (NativeLibraryConfig.LibraryHasLoaded)
+        {
+            // The library is loaded, just pass the callback directly to llama.cpp
+            native_llama_log_set(logCallback);
+
+            // Save a GC handle, to ensure the callback is not collected
+            _currentLogCallbackHandle?.Free();
+            _currentLogCallbackHandle = null;
+            if (logCallback != null)
+                _currentLogCallbackHandle = GCHandle.Alloc(logCallback);
+        }
+        else
+        {
+            // We can't set the log method yet since that would cause the llama.dll to load.
+            // Instead configure it to be set when the native library loading is done
+            NativeLibraryConfig.Instance.WithLogCallback(logCallback);
+        }
+    }
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logger"></param>
+#pragma warning disable IDE1006 // Naming Styles (name imitates llama.cpp)
+    public static void llama_log_set(ILogger? logger)
+#pragma warning restore IDE1006 // Naming Styles
+    {
+        // Clear the logger
+        if (logger == null)
+        {
+            llama_log_set((LLamaLogCallback?)null);
+            return;
+        }
+
+        var builder = new StringBuilder();
+
+        // Bind a function that combines messages until a newline is encountered, then logs it all as one message
+        llama_log_set((level, message) =>
+        {
+            lock (builder)
```

Comment:
There might be a threading problem for example when two models are loading at the same time from different threads. Messages from different models may get mixed up in the builder. That's why I use thread-local approach for builders.

Reply:
I have added this.

Although wouldn't it be a bug in llama.cpp anyway? Since normally these messages go straight to console multiple threads printing at the same time would get interleaved.

Reply:
Yes, you can easily reproduce a mess in console with two threads.
I think logging it is not a priority for llama.cpp, if it is important for the user, they may use llama_log_set and implement something like what we did.